lib/elixir/lib/module/types/descr.ex

```diff
@@ -777,15 +780,113 @@ defmodule Module.Types.Descr do
         end
 
       %{map: map} ->
-        map_split_on_key(map, key)
-        |> Enum.reduce(none(), &union/2)
-        |> pop_optional_static()
+        map_get(map, key) |> pop_optional_static()
 
       %{} ->
         {false, none()}
     end
   end
 
+  @doc """
+  Updates the `key` with a given type, assuming that the key is present
+  in the descr, and that it is exclusively a map (or dynamic).
+  """
+  def map_update(:term, _key, _type), do: :badmap
+  def map_update(descr, key, :term), do: map_update_static_value(descr, key, :term)
+
+  def map_update(descr, key, type) do
+    case :maps.take(:dynamic, type) do
+      :error -> map_update_static_value(descr, key, type)
+      {dynamic, _static} -> dynamic(map_update_static_value(descr, key, dynamic))
+    end
+  end
+
+  def map_update_static_value(descr, key, type) do
+    case :maps.take(:dynamic, descr) do
+      :error ->
+        cond do
+          subtype?(descr, open_map([{key, term()}])) -> map_put_static(descr, key, type)
```

Comment:
Is there a more efficient way of checking this? Perhaps we could use `map_fetch`? Perhaps we could simplify this function with `map_fetch`?